edx_when/api.py

```diff
@@ -152,14 +154,26 @@ def get_dates_for_course(course_id, user=None, use_cached=True, schedule=None):
     external_cache_key = _content_dates_cache_key(course_id, rel_lookup)
     qset = cache.get(external_cache_key) if use_cached else None
     if qset is None:
+        if outline_only:
+            section_usage_key_prefix = str(course_id.make_usage_key('chapter', '_'))[:-1]
+            sequence_usage_key_prefix = str(course_id.make_usage_key('sequential', '_'))[:-1]
+            course_usage_key = course_id.make_usage_key('course', 'course')
+            outline_filter = (
+                Q(location__startswith=section_usage_key_prefix) |
+                Q(location__startswith=sequence_usage_key_prefix) |
+                Q(location=course_usage_key)
+            )
```

Comment:
I think if we're going to allow this kind of filtering, I'd prefer to store the category on the item. It's more work for a migration, though.

I'm also not a huge fan of having to hard-code the types of blocks, but I don't think non-hardcoding works for caching...

Reply:
> I think if we're going to allow this kind of filtering, I'd prefer to store the category on the item. It's more work for a migration, though.

Ah, good point. A lot more robust than my hacky thing here. But also yes to the more work–the migration is simple enough to write, but I don't know how big a yak it is to address perf concerns if we have to actually rebuild the table (or maybe it's null to start, and we lazily update?).

> I'm also not a huge fan of having to hard-code the types of blocks, but I don't think non-hardcoding works for caching...

We could generate a cache key that's dependent on the types of blocks by appending something like `"block_types:{}".format(",".join(sorted(block_types)))` or something.

Reply:
Okay, how about this:

* I create a new, nullable column for `block_type`.
* I put a composite index on `(course_id, block_type)`
* I make it so that writes to this table start filling in values for `block_type` going forward.
* But I don't actually make the data migration that fills it in, so we don't have to worry about locking millions of rows.

In that case, it will only work properly after a new course publish. But the only thing that's going to use it won't work until after a new course publish anyhow.

Reply:
Yeah, that sounds like a good plan. I might also make sure that the query pulls in any rows that don't have a category set (just to make sure we always see the old data, even if we start using the new query pattern).

As far as caching by types: That key would work, I'd just be worried about misses if we start using too many different types... But, on second thought, it's probably resilient enough, and it means we can tailor call-sites as needed.